File: frontend/src/lib/recipe/frontmatter.ts
import { parse as parseYaml } from 'yaml';

export const FENCE_START = '---';
export const FENCE_END = '...';

export interface Frontmatter<M> {
	matter?: M;
	content: string;
}

export function parseFrontmatter<M>(text: string): Frontmatter<M> {
	const lines = splitLines(text);
	const [first, ...rest] = lines;

	if (first === FENCE_START) {
		const end = rest.findIndex((line) => line === FENCE_END);
		if (end > -1) {
			return {
				matter: parseYaml(joinLines(rest.slice(0, end))),
				content: joinLines(rest.slice(end + 1))
			};
		}
	}

	return { content: text };
}

function splitLines(text: string): string[] {
	return text.split('\n');
}

function joinLines(lines: string[]): string {
	return lines.join('\n');
}

File: internal/parser/frontmatter.go
package parser

import (
	"bytes"
	"slices"

	"gopkg.in/yaml.v3"
)

var (
	fenceStart = "---"
	fenceEnd   = "..."
)

type frontmatter[M any] struct {
	Matter  *M
	Content []byte
}

func newFrontmatter[M any](matterBytes []byte, contentBytes []byte) (*frontmatter[M], error) {
	if matterBytes == nil {
		return &frontmatter[M]{Content: contentBytes}, nil
	}

	decoder := yaml.NewDecoder(bytes.NewReader(matterBytes))
	decoder.KnownFields(true)

	var matter M
	if err := decoder.Decode(&matter); err != nil {
		return nil, err
	}

	return &frontmatter[M]{Matter: &matter, Content: contentBytes}, nil
}

func parseFrontmatter[M any](text []byte) (*frontmatter[M], error) {
	lines := splitLines(text)

	if len(lines) > 0 && lines[0] == fenceStart {
		if end := slices.Index(lines, fenceEnd); end > -1 {
			matterBytes := joinLines(lines[1:end])
			contentBytes := joinLines(lines[end+1:])

			return newFrontmatter[M](matterBytes, contentBytes)
		}
	}

	return newFrontmatter[M](nil, text)
}
